File: library/src/main/java/com/goodjia/utility/HidKeyReader.java
package com.goodjia.utility;

import android.os.Handler;
import android.os.Message;
import android.view.KeyEvent;

public class HidKeyReader {
    private static final String TAG = HidKeyReader.class.getSimpleName();
    private static final int HANDLER_WHAT_PUBLISH = 12;
    private static final int DELAY_MILLIS = 300;

    private StringBuilder code = new StringBuilder();
    private Handler handler = new Handler(new Handler.Callback() {
        @Override
        public boolean handleMessage(Message msg) {
            if (msg.what == HANDLER_WHAT_PUBLISH) {
                publish();
            }
            return true;
        }
    });

    private HidKeyListener hidKeyListener;
    private long delayMillis = DELAY_MILLIS;

    public HidKeyReader(HidKeyListener hidKeyListener) {
        this.hidKeyListener = hidKeyListener;
    }

    public void setDelayMillis(long delayMillis) {
        this.delayMillis = delayMillis;
    }

    public void parseKeyEvent(int keyCode, KeyEvent event) {
        if (event.getAction() != KeyEvent.ACTION_DOWN) {
            return;
        }

        if (keyCode == KeyEvent.KEYCODE_ENTER) {
            publish();
        } else {
            code.append((char) event.getUnicodeChar());
            handler.sendEmptyMessageDelayed(HANDLER_WHAT_PUBLISH, delayMillis);
        }
    }

    private void publish() {
        if (!code.toString().isEmpty() && hidKeyListener != null) {
            hidKeyListener.onKeyEvent(code.toString());
        }
        this.code.setLength(0);
        handler.removeCallbacksAndMessages(null);
    }

    public interface HidKeyListener {
        void onKeyEvent(String keyCode);
    }
}

File: library/src/main/java/com/goodjia/utility/HidKeyReader.kt
package com.goodjia.utility

import android.os.Handler
import android.os.HandlerThread
import android.view.KeyEvent
import com.goodjia.utility.HidKeyReader

class HidKeyReader @JvmOverloads constructor(hidKeyListener: HidKeyListener? = null) {
    private val code = StringBuilder()
    private val handlerThread by lazy {
        HandlerThread(TAG)
    }
    private val handler by lazy {
        Handler(handlerThread.looper) { msg ->
            if (msg.what == HANDLER_WHAT_PUBLISH) {
                publish()
            }
            true
        }
    }

    private val hidKeyListeners by lazy {
        mutableListOf<HidKeyListener>()
    }

    var delayMillis = DELAY_MILLIS

    init {
        handlerThread.start()
        hidKeyListener?.let {
            addOnHidKeyListener(it)
        }
    }

    @JvmOverloads
    fun parseKeyEvent(keyCode: Int, event: KeyEvent, action: Int = KeyEvent.ACTION_DOWN) {
        if (event.action != action) return

        if (keyCode == KeyEvent.KEYCODE_ENTER) {
            publish()
        } else {
            code.append(event.unicodeChar.toChar())
            handler.sendEmptyMessageDelayed(HANDLER_WHAT_PUBLISH, delayMillis)
        }
    }

    fun destroy() {
        hidKeyListeners.clear()
        handlerThread.quit()
    }

    fun addOnHidKeyListener(hidKeyListener: HidKeyListener) {
        synchronized(hidKeyListeners) {
            hidKeyListeners.add(hidKeyListener)
        }
    }

    fun removeOnHidKeyListener(hidKeyListener: HidKeyListener) {
        synchronized(hidKeyListeners) {
            hidKeyListeners.remove(hidKeyListener)
        }
    }

    private fun publish() {
        synchronized(hidKeyListeners) {
            if (code.toString().isNotEmpty()) {
                hidKeyListeners.forEach {
                    it.onKeyEvent(code.toString())
                }
            }
            code.setLength(0)
            handler.removeCallbacksAndMessages(null)
        }
    }

    interface HidKeyListener {
        fun onKeyEvent(keyCode: String?)
    }

    companion object {
        private val TAG = HidKeyReader::class.simpleName
        private const val HANDLER_WHAT_PUBLISH = 12
        private const val DELAY_MILLIS = 300L
    }

}